# Importing Data

In [3]:
# Import libraries
import pandas as pd
import numpy as np
import os
path = r'/Users/lamarosavio/Documents/Instacart Basket Analysis'
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)
df_ords = pd.read_csv(os.path.join(path,'02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

# Mixed Type Data

In [7]:
# Create a dataframe
df_test = pd.DataFrame()

In [9]:
#Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [11]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [13]:
# Check for any mixed-type columns
for col in df_test.columns.tolist():
    weird = (df_test[col].map(type) != df_test[col].iloc[0].__class__).any()
    if weird:
        print(col)

mix


In [15]:
# Convert column's data from numeric to string
df_test['mix'] = df_test['mix'].astype('str')

In [17]:
df_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   mix     4 non-null      object
dtypes: object(1)
memory usage: 164.0+ bytes


# Missing Values

In [20]:
# Find missing values (takes sum of Trues in a  column)
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [22]:
# Create subset of the dataframe that contains the nulls
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [24]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

Addressing Missing Values

There a few ways to deal with missing data:
1)  Create a new variable that acts like a flag based on the missing value.
2) Impute the value with the mean or median of the column (if the variable is numeric).
3) Remove or filter out the missing data.

In [30]:
df_nan.describe()

product_id    aisle_id  department_id     prices
count     16.000000   16.000000      16.000000  16.000000
mean    6684.000000   89.937500      10.937500  13.012500
std    12836.665242   33.731229       4.639953   3.881731
min       34.000000   26.000000       1.000000   1.200000
25%      459.250000   70.750000       7.750000  12.175000
50%     2413.000000   98.500000      11.500000  13.650000
75%     3872.750000  120.000000      14.500000  14.425000
max    40440.000000  126.000000      16.000000  20.900000

Ex: When using the mean, df['column with missings'].fillna(mean value, inplace=True)

In [33]:
df_nan.median()

product_id       2413.0
product_name        NaN
aisle_id           98.5
department_id      11.5
prices            13.65
dtype: object

Because the missing values are strings there's not much you can do other than remove/filter the data

In [36]:
# Find shape of df
df_prods.shape

(49693, 5)

In [38]:
# Create new df without the nulls
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [40]:
# Should have exactly 16 less rows
df_prods_clean.shape

(49677, 5)

To drop all missing values --> df_prods.dropna(inplace = True)
To drop only the NaNs from a particular column --> df_prods.dropna(subset = [‘product_name’], inplace = True)

# Duplicates

In [44]:
# Create subset of df_prods_clean that contains only rows of duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [46]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [48]:
# Check current number of rowns in df_prods_clean
df_prods_clean.shape

(49677, 5)

In [50]:
# Create datafram that doesn't include the duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [52]:
df_prods_clean_no_dups.shape

(49672, 5)

# Exporting Changes

In [55]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# Answers 4.5

Step 2

In [61]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [63]:
# Import libraries
import pandas as pd
import numpy as np
import os
path = r'/Users/lamarosavio/Documents/Instacart Basket Analysis'
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)
df_ords = pd.read_csv(os.path.join(path,'02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [65]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

Step 3

In [68]:
for col in df_ords.columns.tolist():
    weird = (df_ords[col].map(type) != df_ords[col].iloc[0].__class__).any()
    if weird:
        print(col)

Unnamed: 0
order_id
user_id
order_number
order_dow
order_hour_of_day
days_since_prior_order


In [70]:
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   Unnamed: 0              int64  
 1   order_id                int64  
 2   user_id                 int64  
 3   order_number            int64  
 4   order_dow               int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(6)
memory usage: 182.7 MB


Step 5

In [73]:
# Find missing values (takes sum of Trues in a  column)
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

days_since_prior_order is missing 206209 values

Step 6

In [77]:
# Create subset of the dataframe that contains the nulls
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [79]:
df_ords_nan

Unnamed: 0  order_id  user_id  order_number  order_dow  \
0                 0   2539329        1             1          2   
11               11   2168274        2             1          2   
26               26   1374495        3             1          1   
39               39   3343014        4             1          6   
45               45   2717275        5             1          3   
...             ...       ...      ...           ...        ...   
3420930     3420930    969311   206205             1          4   
3420934     3420934   3189322   206206             1          3   
3421002     3421002   2166133   206207             1          6   
3421019     3421019   2227043   206208             1          1   
3421069     3421069   3154581   206209             1          3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

In [83]:
df_ords.median()

Unnamed: 0                1710541.0
order_id                  1710542.0
user_id                    102689.0
order_number                   11.0
order_dow                       3.0
order_hour_of_day              13.0
days_since_prior_order          7.0
dtype: float64

In [91]:
df_ords[df_ords['order_number']==1]

Unnamed: 0  order_id  user_id  order_number  order_dow  \
0                 0   2539329        1             1          2   
11               11   2168274        2             1          2   
26               26   1374495        3             1          1   
39               39   3343014        4             1          6   
45               45   2717275        5             1          3   
...             ...       ...      ...           ...        ...   
3420930     3420930    969311   206205             1          4   
3420934     3420934   3189322   206206             1          3   
3421002     3421002   2166133   206207             1          6   
3421019     3421019   2227043   206208             1          1   
3421069     3421069   3154581   206209             1          3   

         order_hour_of_day  days_since_prior_order  
0                        8                     7.0  
11                      11                     7.0  
26                      14                     7.0  
39                      11                     7.0  
45                      12                     7.0  
...                    ...                     ...  
3420930                 12                     7.0  
3420934                 18                     7.0  
3421002                 19                     7.0  
3421019                 15                     7.0  
3421069                 11                     7.0  

[206209 rows x 7 columns]

Step 7

In [94]:
# Create subset of df_ords that contains only rows of duplicates
df_ords_dups = df_ords[df_ords.duplicated()]

In [96]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

There are no duplicates

In [98]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))